File: packages/alpaca-option/src/marketStructure.ts
import { clock as timeClock } from '@alpaca/time';

import type {
  MarketStructureAnalysis,
  MarketStructureFilters,
  MarketStructureLevel,
  MarketStructureOptionRecord,
} from './types';

type LevelAccumulator = {
  strike: number;
  call_open_interest: number;
  put_open_interest: number;
  call_gamma_exposure: number;
  put_gamma_exposure: number;
  call_volume: number;
  put_volume: number;
};

export function gammaExposure(
  record: MarketStructureOptionRecord,
  underlyingPrice: number,
): number | null {
  const gamma = finite(record.gamma);
  const openInterest = finite(record.open_interest);
  const multiplier = finite(record.multiplier);
  const spot = finite(underlyingPrice);

  if (
    gamma == null ||
    openInterest == null ||
    multiplier == null ||
    spot == null ||
    openInterest < 0 ||
    multiplier <= 0 ||
    spot <= 0
  ) {
    return null;
  }

  const exposure = gamma * openInterest * multiplier * spot * spot * 0.01;
  return finite(record.option_right === 'put' ? -exposure : exposure);
}

export function filterMarketStructureRecords(
  records: MarketStructureOptionRecord[],
  filters: MarketStructureFilters,
): MarketStructureOptionRecord[] {
  return records.filter(
    (record) =>
      matchesExpiration(record, filters) &&
      matchesOptionRight(record, filters) &&
      matchesStrike(record, filters) &&
      matchesDte(record, filters) &&
      matchesOpenInterest(record, filters),
  );
}

export function analyzeMarketStructure(
  records: MarketStructureOptionRecord[],
): MarketStructureAnalysis {
  const recordsCount = records.length;
  const underlyingPrice =
    records
      .map((record) => finitePositive(record.underlying_price))
      .find((value) => value != null) ?? null;
  const openInterestCount = records.filter(
    (record) => finite(record.open_interest) != null,
  ).length;
  const warnings: string[] = [];

  if (records.length === 0) {
    warnings.push('no_records');
  }
  if (underlyingPrice == null) {
    warnings.push('missing_underlying_price');
  }
  if (openInterestCount < recordsCount) {
    warnings.push('incomplete_open_interest');
  }

  const accumulators: LevelAccumulator[] = [];
  if (underlyingPrice != null) {
    for (const record of records) {
      const strike = finite(record.strike);
      if (strike == null) {
        continue;
      }

      const openInterest = Math.max(finite(record.open_interest) ?? 0, 0);
      const volume = activityVolume(record);
      const exposure = gammaExposure(record, underlyingPrice) ?? 0;
      let level = accumulators.find((candidate) => sameStrike(candidate.strike, strike));
      if (!level) {
        level = {
          strike,
          call_open_interest: 0,
          put_open_interest: 0,
          call_gamma_exposure: 0,
          put_gamma_exposure: 0,
          call_volume: 0,
          put_volume: 0,
        };
        accumulators.push(level);
      }

      if (record.option_right === 'call') {
        level.call_open_interest += openInterest;
        level.call_gamma_exposure += exposure;
        level.call_volume += volume;
      } else {
        level.put_open_interest += openInterest;
        level.put_gamma_exposure += exposure;
        level.put_volume += volume;
      }
    }
  }

  const levels = accumulators
    .sort((left, right) => left.strike - right.strike)
    .map((level) => toLevel(level, []));

  if (levels.length === 0 && recordsCount > 0) {
    warnings.push('no_gamma_exposure_records');
  }

  const callWallStrike =
    maxBy(
      levels.filter((level) => level.call_gamma_exposure > 0),
      (level) => level.call_gamma_exposure,
    )?.strike ?? null;
  const putWallStrike =
    minBy(
      levels.filter((level) => level.put_gamma_exposure < 0),
      (level) => level.put_gamma_exposure,
    )?.strike ?? null;
  const absoluteWallStrike =
    maxBy(
      levels.filter((level) => level.absolute_gamma_exposure > 0),
      (level) => level.absolute_gamma_exposure,
    )?.strike ?? null;

  for (const level of levels) {
    if (callWallStrike != null && sameStrike(level.strike, callWallStrike)) {
      level.labels.push('call_wall');
    }
    if (putWallStrike != null && sameStrike(level.strike, putWallStrike)) {
      level.labels.push('put_wall');
    }
    if (absoluteWallStrike != null && sameStrike(level.strike, absoluteWallStrike)) {
      level.labels.push('absolute_gamma_exposure_wall');
    }
  }

  const netGammaExposure = levels.reduce((sum, level) => sum + level.net_gamma_exposure, 0);
  const absoluteGammaExposure = levels.reduce(
    (sum, level) => sum + level.absolute_gamma_exposure,
    0,
  );

  return {
    underlying_price: underlyingPrice,
    records_count: recordsCount,
    open_interest_coverage: recordsCount === 0 ? 0 : openInterestCount / recordsCount,
    call_wall: levelByStrike(levels, callWallStrike),
    put_wall: levelByStrike(levels, putWallStrike),
    absolute_gamma_exposure_wall: levelByStrike(levels, absoluteWallStrike),
    net_gamma_exposure: netGammaExposure,
    absolute_gamma_exposure: absoluteGammaExposure,
    levels,
    warnings,
  };
}

function toLevel(level: LevelAccumulator, labels: string[]): MarketStructureLevel {
  const totalOpenInterest = level.call_open_interest + level.put_open_interest;
  const netGammaExposure = level.call_gamma_exposure + level.put_gamma_exposure;
  const absoluteGammaExposure =
    Math.abs(level.call_gamma_exposure) + Math.abs(level.put_gamma_exposure);
  const totalVolume = level.call_volume + level.put_volume;

  return {
    strike: level.strike,
    call_open_interest: level.call_open_interest,
    put_open_interest: level.put_open_interest,
    total_open_interest: totalOpenInterest,
    call_gamma_exposure: level.call_gamma_exposure,
    put_gamma_exposure: level.put_gamma_exposure,
    net_gamma_exposure: netGammaExposure,
    absolute_gamma_exposure: absoluteGammaExposure,
    call_volume: level.call_volume,
    put_volume: level.put_volume,
    total_volume: totalVolume,
    labels,
  };
}

function matchesExpiration(
  record: MarketStructureOptionRecord,
  filters: MarketStructureFilters,
): boolean {
  return !filters.expiration_date || record.expiration_date === filters.expiration_date;
}

function matchesOptionRight(
  record: MarketStructureOptionRecord,
  filters: MarketStructureFilters,
): boolean {
  return !filters.option_right || record.option_right === filters.option_right;
}

function matchesStrike(
  record: MarketStructureOptionRecord,
  filters: MarketStructureFilters,
): boolean {
  const strike = finite(record.strike);
  if (strike == null) {
    return false;
  }

  const min = finite(filters.strike_price_gte);
  if (min != null && strike < min) {
    return false;
  }
  const max = finite(filters.strike_price_lte);
  if (max != null && strike > max) {
    return false;
  }
  return true;
}

function matchesDte(
  record: MarketStructureOptionRecord,
  filters: MarketStructureFilters,
): boolean {
  if (filters.dte_min == null && filters.dte_max == null) {
    return true;
  }

  const dte = daysToExpiration(record);
  if (dte == null) {
    return false;
  }
  const min = finite(filters.dte_min);
  if (min != null && dte < min) {
    return false;
  }
  const max = finite(filters.dte_max);
  if (max != null && dte > max) {
    return false;
  }
  return true;
}

function matchesOpenInterest(
  record: MarketStructureOptionRecord,
  filters: MarketStructureFilters,
): boolean {
  if (!filters.require_open_interest) {
    return true;
  }
  const openInterest = finite(record.open_interest);
  return openInterest != null && openInterest > 0;
}

function daysToExpiration(record: MarketStructureOptionRecord): number | null {
  const asOfDate = record.as_of.slice(0, 10) || record.as_of;
  return finite(timeClock.fractionalDaysBetween(asOfDate, record.expiration_date));
}

function activityVolume(record: MarketStructureOptionRecord): number {
  return record.daily_volume ?? record.minute_volume ?? record.latest_trade_size ?? 0;
}

function levelByStrike(
  levels: MarketStructureLevel[],
  strike: number | null,
): MarketStructureLevel | null {
  if (strike == null) {
    return null;
  }
  return levels.find((level) => sameStrike(level.strike, strike)) ?? null;
}

function finite(value: number | null | undefined): number | null {
  return typeof value === 'number' && Number.isFinite(value) ? value : null;
}

function finitePositive(value: number | null | undefined): number | null {
  const number = finite(value);
  return number != null && number > 0 ? number : null;
}

function sameStrike(left: number, right: number): boolean {
  return Math.abs(left - right) <= 1e-8;
}

function maxBy<T>(values: T[], score: (value: T) => number): T | undefined {
  let best: T | undefined;
  let bestScore = Number.NEGATIVE_INFINITY;
  for (const value of values) {
    const current = score(value);
    if (current > bestScore) {
      best = value;
      bestScore = current;
    }
  }
  return best;
}

function minBy<T>(values: T[], score: (value: T) => number): T | undefined {
  let best: T | undefined;
  let bestScore = Number.POSITIVE_INFINITY;
  for (const value of values) {
    const current = score(value);
    if (current < bestScore) {
      best = value;
      bestScore = current;
    }
  }
  return best;
}

File: crates/alpaca-option/src/market_structure.rs
use std::cmp::Ordering;

use alpaca_time::clock;

use crate::types::{
    MarketStructureAnalysis, MarketStructureAnalysisOptions, MarketStructureExposureMode,
    MarketStructureFilters, MarketStructureLevel, MarketStructureOptionRecord, OptionRight,
};

#[derive(Debug, Clone, Default)]
struct LevelAccumulator {
    strike: f64,
    call_open_interest: f64,
    put_open_interest: f64,
    call_gamma_exposure: f64,
    put_gamma_exposure: f64,
    call_volume: u64,
    put_volume: u64,
}

impl LevelAccumulator {
    fn into_level(self, labels: Vec<String>) -> MarketStructureLevel {
        let total_open_interest = self.call_open_interest + self.put_open_interest;
        let net_gamma_exposure = self.call_gamma_exposure + self.put_gamma_exposure;
        let absolute_gamma_exposure =
            self.call_gamma_exposure.abs() + self.put_gamma_exposure.abs();
        let total_volume = self.call_volume.saturating_add(self.put_volume);

        MarketStructureLevel {
            strike: self.strike,
            call_open_interest: self.call_open_interest,
            put_open_interest: self.put_open_interest,
            total_open_interest,
            call_gamma_exposure: self.call_gamma_exposure,
            put_gamma_exposure: self.put_gamma_exposure,
            net_gamma_exposure,
            absolute_gamma_exposure,
            call_volume: self.call_volume,
            put_volume: self.put_volume,
            total_volume,
            labels,
        }
    }
}

pub fn gamma_exposure(record: &MarketStructureOptionRecord, underlying_price: f64) -> Option<f64> {
    gamma_exposure_with_mode(
        record,
        underlying_price,
        MarketStructureExposureMode::GexProxy,
    )
}

pub fn gamma_exposure_with_mode(
    record: &MarketStructureOptionRecord,
    underlying_price: f64,
    mode: MarketStructureExposureMode,
) -> Option<f64> {
    gamma_exposure_from_gamma(record, underlying_price, record.gamma?, mode)
}

pub fn filter_market_structure_records(
    records: &[MarketStructureOptionRecord],
    filters: &MarketStructureFilters,
) -> Vec<MarketStructureOptionRecord> {
    records
        .iter()
        .filter(|record| matches_expiration(record, filters))
        .filter(|record| matches_option_right(record, filters))
        .filter(|record| matches_strike(record, filters))
        .filter(|record| matches_dte(record, filters))
        .filter(|record| matches_open_interest(record, filters))
        .cloned()
        .collect()
}

pub fn analyze_market_structure(
    records: &[MarketStructureOptionRecord],
) -> MarketStructureAnalysis {
    analyze_market_structure_with_options(records, &MarketStructureAnalysisOptions::default())
}

pub fn analyze_market_structure_with_options(
    records: &[MarketStructureOptionRecord],
    options: &MarketStructureAnalysisOptions,
) -> MarketStructureAnalysis {
    let records_count = records.len();
    let underlying_price = records
        .iter()
        .filter_map(|record| record.underlying_price.and_then(finite_positive))
        .next();
    let open_interest_count = records
        .iter()
        .filter(|record| record.open_interest.and_then(finite).is_some())
        .count();

    let mut warnings = Vec::new();
    if records.is_empty() {
        warnings.push("no_records".to_string());
    }
    if underlying_price.is_none() {
        warnings.push("missing_underlying_price".to_string());
    }
    if open_interest_count < records_count {
        warnings.push("incomplete_open_interest".to_string());
    }

    let mut accumulators = Vec::<LevelAccumulator>::new();
    if let Some(spot) = underlying_price {
        for record in records {
            let strike = match finite(record.strike) {
                Some(value) => value,
                None => continue,
            };
            let open_interest = record
                .open_interest
                .and_then(finite)
                .unwrap_or(0.0)
                .max(0.0);
            let volume = activity_volume(record);
            let exposure = gamma_exposure_with_mode(record, spot, options.mode).unwrap_or(0.0);

            let index = accumulators
                .iter()
                .position(|level| same_strike(level.strike, strike))
                .unwrap_or_else(|| {
                    accumulators.push(LevelAccumulator {
                        strike,
                        ..LevelAccumulator::default()
                    });
                    accumulators.len() - 1
                });
            let level = &mut accumulators[index];

            match record.option_right {
                OptionRight::Call => {
                    level.call_open_interest += open_interest;
                    level.call_gamma_exposure += exposure;
                    level.call_volume = level.call_volume.saturating_add(volume);
                }
                OptionRight::Put => {
                    level.put_open_interest += open_interest;
                    level.put_gamma_exposure += exposure;
                    level.put_volume = level.put_volume.saturating_add(volume);
                }
            }
        }
    }

    accumulators.sort_by(|left, right| compare_f64(left.strike, right.strike));

    let mut levels: Vec<MarketStructureLevel> = accumulators
        .into_iter()
        .map(|level| level.into_level(Vec::new()))
        .collect();

    if records_count > 0
        && (levels.is_empty()
            || !levels
                .iter()
                .any(|level| level.absolute_gamma_exposure > 0.0))
    {
        warnings.push("no_gamma_exposure_records".to_string());
    }

    let call_wall_strike = levels
        .iter()
        .filter(|level| level.call_gamma_exposure != 0.0)
        .max_by(|left, right| {
            compare_f64(
                left.call_gamma_exposure.abs(),
                right.call_gamma_exposure.abs(),
            )
        })
        .map(|level| level.strike)
        .or_else(|| {
            levels
                .iter()
                .filter(|level| level.call_open_interest > 0.0)
                .max_by(|left, right| {
                    compare_f64(left.call_open_interest, right.call_open_interest)
                })
                .map(|level| level.strike)
        });
    let put_wall_strike = levels
        .iter()
        .filter(|level| level.put_gamma_exposure != 0.0)
        .max_by(|left, right| {
            compare_f64(
                left.put_gamma_exposure.abs(),
                right.put_gamma_exposure.abs(),
            )
        })
        .map(|level| level.strike)
        .or_else(|| {
            levels
                .iter()
                .filter(|level| level.put_open_interest > 0.0)
                .max_by(|left, right| compare_f64(left.put_open_interest, right.put_open_interest))
                .map(|level| level.strike)
        });
    let absolute_wall_strike = levels
        .iter()
        .filter(|level| level.absolute_gamma_exposure > 0.0)
        .max_by(|left, right| {
            compare_f64(left.absolute_gamma_exposure, right.absolute_gamma_exposure)
        })
        .map(|level| level.strike)
        .or_else(|| {
            levels
                .iter()
                .filter(|level| level.total_open_interest > 0.0)
                .max_by(|left, right| {
                    compare_f64(left.total_open_interest, right.total_open_interest)
                })
                .map(|level| level.strike)
        });

    for level in &mut levels {
        if call_wall_strike
            .map(|strike| same_strike(level.strike, strike))
            .unwrap_or(false)
        {
            level.labels.push("call_wall".to_string());
        }
        if put_wall_strike
            .map(|strike| same_strike(level.strike, strike))
            .unwrap_or(false)
        {
            level.labels.push("put_wall".to_string());
        }
        if absolute_wall_strike
            .map(|strike| same_strike(level.strike, strike))
            .unwrap_or(false)
        {
            level
                .labels
                .push("absolute_gamma_exposure_wall".to_string());
        }
    }

    let call_wall = level_by_strike(&levels, call_wall_strike);
    let put_wall = level_by_strike(&levels, put_wall_strike);
    let absolute_gamma_exposure_wall = level_by_strike(&levels, absolute_wall_strike);
    let net_gamma_exposure = levels.iter().map(|level| level.net_gamma_exposure).sum();
    let absolute_gamma_exposure = levels
        .iter()
        .map(|level| level.absolute_gamma_exposure)
        .sum();
    let open_interest_coverage = if records_count == 0 {
        0.0
    } else {
        open_interest_count as f64 / records_count as f64
    };

    MarketStructureAnalysis {
        underlying_price,
        records_count,
        open_interest_coverage,
        call_wall,
        put_wall,
        absolute_gamma_exposure_wall,
        net_gamma_exposure,
        absolute_gamma_exposure,
        levels,
        warnings,
    }
}

fn matches_expiration(
    record: &MarketStructureOptionRecord,
    filters: &MarketStructureFilters,
) -> bool {
    filters
        .expiration_date
        .as_ref()
        .map(|expiration| record.expiration_date == *expiration)
        .unwrap_or(true)
}

fn matches_option_right(
    record: &MarketStructureOptionRecord,
    filters: &MarketStructureFilters,
) -> bool {
    filters
        .option_right
        .as_ref()
        .map(|right| record.option_right == *right)
        .unwrap_or(true)
}

fn matches_strike(record: &MarketStructureOptionRecord, filters: &MarketStructureFilters) -> bool {
    let Some(strike) = finite(record.strike) else {
        return false;
    };

    if filters
        .strike_price_gte
        .and_then(finite)
        .map(|min| strike < min)
        .unwrap_or(false)
    {
        return false;
    }
    if filters
        .strike_price_lte
        .and_then(finite)
        .map(|max| strike > max)
        .unwrap_or(false)
    {
        return false;
    }
    true
}

fn matches_dte(record: &MarketStructureOptionRecord, filters: &MarketStructureFilters) -> bool {
    if filters.dte_min.is_none() && filters.dte_max.is_none() {
        return true;
    }

    let Some(dte) = days_to_expiration(record) else {
        return false;
    };
    if filters
        .dte_min
        .and_then(finite)
        .map(|min| dte < min)
        .unwrap_or(false)
    {
        return false;
    }
    if filters
        .dte_max
        .and_then(finite)
        .map(|max| dte > max)
        .unwrap_or(false)
    {
        return false;
    }
    true
}

fn matches_open_interest(
    record: &MarketStructureOptionRecord,
    filters: &MarketStructureFilters,
) -> bool {
    if !filters.require_open_interest {
        return true;
    }
    record
        .open_interest
        .and_then(finite)
        .map(|open_interest| open_interest > 0.0)
        .unwrap_or(false)
}

fn days_to_expiration(record: &MarketStructureOptionRecord) -> Option<f64> {
    let as_of_date = record.as_of.get(0..10).unwrap_or(&record.as_of);
    clock::fractional_days_between(as_of_date, &record.expiration_date)
        .ok()
        .and_then(finite)
}

fn activity_volume(record: &MarketStructureOptionRecord) -> u64 {
    record
        .daily_volume
        .or(record.minute_volume)
        .or(record.latest_trade_size)
        .unwrap_or(0)
}

fn level_by_strike(
    levels: &[MarketStructureLevel],
    strike: Option<f64>,
) -> Option<MarketStructureLevel> {
    let strike = strike?;
    levels
        .iter()
        .find(|level| same_strike(level.strike, strike))
        .cloned()
}

fn gamma_exposure_from_gamma(
    record: &MarketStructureOptionRecord,
    spot: f64,
    gamma: f64,
    mode: MarketStructureExposureMode,
) -> Option<f64> {
    let gamma = finite(gamma)?;
    let open_interest = finite(record.open_interest?)?;
    let multiplier = finite(record.multiplier?)?;
    let spot = finite(spot)?;
    if open_interest < 0.0 || multiplier <= 0.0 || spot <= 0.0 {
        return None;
    }

    let exposure = gamma * open_interest * multiplier * spot * spot * 0.01;
    let signed_exposure = match (mode, record.option_right.clone()) {
        (MarketStructureExposureMode::GexProxy, OptionRight::Call)
        | (MarketStructureExposureMode::DealerView, OptionRight::Put) => exposure,
        (MarketStructureExposureMode::GexProxy, OptionRight::Put)
        | (MarketStructureExposureMode::DealerView, OptionRight::Call) => -exposure,
    };
    finite(signed_exposure)
}

fn finite(value: f64) -> Option<f64> {
    if value.is_finite() {
        Some(value)
    } else {
        None
    }
}

fn finite_positive(value: f64) -> Option<f64> {
    finite(value).filter(|value| *value > 0.0)
}

fn same_strike(left: f64, right: f64) -> bool {
    (left - right).abs() <= 1e-8
}

fn compare_f64(left: f64, right: f64) -> Ordering {
    left.partial_cmp(&right).unwrap_or(Ordering::Equal)
}
